content/3_Silver/Graph_Traversal.mdx

```diff
@@ -247,6 +166,286 @@ d.pop()  # [1, 3]
+	for neighbor in adj[current_node]:
+		if not visited[neighbor]:
+			dfs(neighbor)
+
+
+for i in range(MAX_N):
+	if not visited[i]:
+		dfs(i)
+```
+
+</PySection>
+<JavaSection>
+
+```java
+static ArrayList<ArrayList<Integer>> adj;
+static boolean[] visited;
+static final int MAX_N = (int)1e6;
+
+public static void main(String[] args) throws IOException {
+	visited = new boolean[MAX_N];
+	adj = new ArrayList<ArrayList<Integer>>();
+	for (int i = 0; i < MAX_N; i++) adj.add(new ArrayList<Integer>());
+
+	// Define adjacency list and read in problem-specific input
+
+	for (int i = 0; i < MAX_N; i++) {
+		// iterate over all connected components in the graph
+		if (!visited[i]) { dfs(i); }
+	}
+}
+
+static void dfs(int currentNode) {
+	if (visited[currentNode]) { return; }
+
+	visited[currentNode] = true;
+	for (int neighbor : adj.get(currentNode)) { dfs(neighbor); }
+}
+```
+
+</JavaSection>
+<CPPSection>
+
+```cpp
+const int MAX_N = 1e6;
+vector<int> adj[MAX_N];
+bool visited[MAX_N];
+
+void dfs(int current_node) {
+	if (visited[current_node]) { return; }
+	visited[current_node] = true;
+
+	for (int neighbor : adj[current_node]) { dfs(neighbor); }
+}
+
+int main() {
+	// Define adjacency list and read in problem-specific input
+
+	for (int i = 0; i < MAX_N; i++) {
+		// iterate over all connected components in the graph
+		if (!visited[i]) { dfs(i); }
+	}
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+
+#### BFS
+
+When implementing BFS, we often use a queue to track the next vertex to visit. Like DFS, we'll also keep an array to store whether we've seen a vertex before.
+
+The code below shows the variables we need to maintain and an iterative implementation of DFS.
+
+<LanguageSection>
+<JavaSection>
+
+```java
+static final int MAX_N = (int)1e6;
+
+public static void main(String[] args) throws IOException {
+	boolean[] visited = new boolean[MAX_N];
+	ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
+	for (int i = 0; i < MAX_N; i++) adj.add(new ArrayList<Integer>());
+
+	// Define adjacency list and read in problem-specific input
+
+	for (int i = 0; i < MAX_N; i++) {
+		// iterate over all connected components in the graph
+		if (!visited[i]) {
+			Queue<Integer> q = new LinkedList<Integer>();
+			q.add(i);
+
+			while (q.size() > 0) {
+				int currentNode = q.peek();
+				q.poll();
+
+				if (visited[currentNode]) { continue; }
+				visited[currentNode] = true;
+
+				for (int neighbor : adj.get(currentNode)) { q.add(neighbor); }
+			}
+		}
+	}
+}
+```
+
+</JavaSection>
+<CPPSection>
+
+```cpp
+const int MAX_N = 1e6;
+
+int main() {
+	vector<int> adj[MAX_N];
+	bool visited[MAX_N] = {0};
+
+	// Define adjacency list and read in problem-specific input
+
+	for (int i = 0; i < MAX_N; i++) {
+		// iterate over all connected components in the graph
+		if (!visited[i]) {
+			queue<int> q;
+			q.push(i);
+
+			while (!q.empty()) {
+				int current_node = q.front();
+				q.pop();
+
+				if (visited[current_node]) { continue; }
```

Comment:
i suggest checking for visited nodes inside the neighbors loop, in some instances, it's a pretty significant speed build up (since youre constantly pushing already visited nodes in the queue)